multi/imxrt-multi/cm4.c

```diff
@@ -279,127 +316,84 @@ static int loadFromBuff(const void *buff, size_t bufflen)
 
 static void devctl(msg_t *msg)
 {
-	imxrt117xM4DevCtli_t *i = (void *)msg->i.raw;
-	imxrt117xM4DevCtlo_t *o = (void *)msg->o.raw;
+	multi_i_t *iptr = (multi_i_t *)msg->i.raw;
+	multi_o_t *optr = (multi_o_t *)msg->o.raw;
+
+	switch (iptr->cm4_type) {
+		case CM4_LOAD_BUFF:
+			if ((msg->i.data == NULL) || (msg->i.size == 0)) {
+				optr->err = -EINVAL;
+				break;
+			}
 
-	switch (i->type) {
-	case m4_loadBuff:
-		if (msg->i.data == NULL || msg->i.size == 0) {
-			o->err = -EINVAL;
+			optr->err = loadFromBuff(msg->i.data, msg->i.size);
 			break;
-		}
 
-		o->err = loadFromBuff(msg->i.data, msg->i.size);
-		break;
+		case CM4_LOAD_FILE:
+			if ((msg->i.data == NULL) || (msg->i.size == 0)) {
+				optr->err = -EINVAL;
+				break;
+			}
 
-	case m4_loadFile:
-		if (msg->i.data == NULL || msg->i.size == 0) {
-			o->err = -EINVAL;
+			optr->err = loadFromFile(msg->i.data);
 			break;
-		}
-
-		o->err = loadFromFile(msg->i.data);
-		break;
 
-	case m4_runCore:
-		if (msg->i.data == NULL || msg->i.size != sizeof(unsigned int)) {
-			o->err = -EINVAL;
+		case CM4_RUN_CORE:
+			if ((msg->i.data == NULL) || (msg->i.size != sizeof(unsigned int))) {
+				optr->err = -EINVAL;
+				break;
+			}
+			optr->err = runCore(*(unsigned int *)msg->i.data);
 			break;
-		}
-		o->err = runCore(*(unsigned int *)msg->i.data);
-		break;
 
-	default:
-		o->err = -EINVAL;
-		break;
+		default:
+			optr->err = -EINVAL;
+			break;
 	}
 }
 
 
-__attribute__((noreturn)) static void thread(void)
+void cm4_handleMsg(msg_t *msg)
 {
-	msg_t msg;
-	unsigned long rid;
-
-	for (;;) {
-		if (msgRecv(m4_common.port, &msg, &rid) < 0)
-			continue;
-
-		switch (msg.type) {
+	mutexLock(m4_common.lock);
+	switch (msg->type) {
 		case mtRead:
-			msg.o.io.err = chanRead(msg.i.io.oid.id, msg.o.data, msg.o.size);
+			msg->o.io.err = chanRead(msg->i.io.oid.id - id_cm4_0, msg->o.data, msg->o.size);
 			break;
 
 		case mtWrite:
-			msg.o.io.err = chanWrite(msg.i.io.oid.id, msg.i.data, msg.i.size);
+			msg->o.io.err = chanWrite(msg->i.io.oid.id - id_cm4_0, msg->i.data, msg->i.size);
 			break;
 
-		case mtDevCtl:
-			devctl(&msg);
+		case mtGetAttr:
+		case mtSetAttr:
+			msg->o.attr.err = -ENOSYS;
 			break;
 
-		case mtOpen:
-		case mtClose:
-			if (msg.i.openclose.oid.port == m4_common.port && msg.i.openclose.oid.id >= chan0 && msg.i.openclose.oid.id < chanLen)
-				msg.o.io.err = EOK;
-			else
-				msg.o.io.err = -EINVAL;
+		case mtDevCtl:
+			devctl(msg);
 			break;
 
-		/* Unsupported operations */
-		case mtCreate:
-			msg.o.create.err = -ENOSYS;
-			break;
-		case mtLookup:
-			msg.o.lookup.err = -ENOSYS;
-			break;
 		default:
-			msg.o.io.err = -ENOSYS;
+			msg->o.io.err = -EINVAL;
```

Comment:
Why changed to `EINVAL` (Invalid argument) ?
`ENOSYS` a.k.a function not implemented, for me suits better. Other handlers in imxrt-multi default err is ENOSYS, see gpio, spi.